File: leetcode/0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.js
/**
 * @param {number[]} heights
 * @return {number}
 */
function largestRectangleArea(heights) {
    const stack = [];
    let result = 0;

    for (let idx = 0; idx < heights.length; idx++) {
        const height = heights[idx];

        if (stack.length === 0) {
            stack.push([idx, height]);
        } else {
            let width = idx;

            while (
                stack.length > 0 &&
                stack[stack.length - 1][1] > height
            ) {
                const value = stack.pop();
                width = value[0];
                const size = value[1] * (idx - value[0]);
                result = Math.max(result, size);
            }

            stack.push([width, height]);
        }
    }

    while (stack.length > 0) {
        const value = stack.shift();
        const size = value[1] * (heights.length - value[0]);
        result = Math.max(result, size);
    }

    return result;
}

File: leetcode/0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Deque<int[]> stack = new ArrayDeque<>();
        int result = 0;

        for (int idx = 0; idx < heights.length; idx++) {
            int height = heights[idx];

            if (stack.isEmpty()) {
                stack.push(new int[]{idx, height});
            } else {
                int width = idx;

                while (!stack.isEmpty() && stack.peek()[1] > height) {
                    int[] value = stack.pop();
                    width = value[0];
                    int size = value[1] * (idx - value[0]);
                    result = Math.max(result, size);
                }

                stack.push(new int[]{width, height});
            }
        }

        while (!stack.isEmpty()) {
            int[] value = stack.removeLast();
            int size = value[1] * (heights.length - value[0]);
            result = Math.max(result, size);
        }

        return result;
    }
}